src/session/content/message_row/sticker.rs

```diff
@@ -48,48 +44,48 @@ mod imp {
+    pub fn new(message: &glib::Object) -> Self {
+        glib::Object::new(&[("message", message)]).expect("Failed to create MessageSticker")
     }
 
-    pub fn set_message(&self, message: &Message) {
-        if let MessageContent::MessageSticker(data) = message.content().0 {
-            let self_ = imp::MessageSticker::from_instance(self);
-            self_
-                .paintable
-                .set_aspect_ratio(data.sticker.width as f64 / data.sticker.height as f64);
-
-            if data.sticker.sticker.local.is_downloading_completed {
-                self.load_sticker(&data.sticker.sticker.local.path);
+    fn update_widget(&self) {
+        if let Some(message) = self.message() {
+            if let Some(message) = message.downcast_ref::<Message>() {
+                if let MessageContent::MessageSticker(data) = message.content().0 {
+                    let self_ = imp::MessageSticker::from_instance(self);
+                    self_
+                        .paintable
+                        .set_aspect_ratio(data.sticker.width as f64 / data.sticker.height as f64);
+
+                    if data.sticker.sticker.local.is_downloading_completed {
+                        self.load_sticker(&data.sticker.sticker.local.path);
+                    } else {
+                        let (sender, receiver) =
+                            glib::MainContext::sync_channel::<File>(Default::default(), 5);
+
+                        receiver.attach(
+                            None,
+                            clone!(@weak self as obj => @default-return glib::Continue(false), move |file| {
+                                if file.local.is_downloading_completed {
+                                    obj.load_sticker(&file.local.path);
+                                }
+
+                                glib::Continue(true)
+                            }),
+                        );
+
+                        message
+                            .chat()
+                            .session()
+                            .download_file(data.sticker.sticker.id, sender);
+                    }
+                }
             } else {
-                let (sender, receiver) =
-                    glib::MainContext::sync_channel::<File>(Default::default(), 5);
-
-                receiver.attach(
-                    None,
-                    clone!(@weak self as obj => @default-return glib::Continue(false), move |file| {
-                        if file.local.is_downloading_completed {
-                            obj.load_sticker(&file.local.path);
-                        }
-
-                        glib::Continue(true)
-                    }),
-                );
-
-                message
-                    .chat()
-                    .session()
-                    .download_file(data.sticker.sticker.id, sender);
+                log::warn!("Unexpected message type: {:?}", message);
```

Comment:
Not sure whether just printing a warning is the correct way here.
As I understand, we are responsible for ensuring that this sticker's message object is actually a `Message` and not a `SponsoredMessage` (These both types are set in `MessageRowExt::set_message()`).
So just logging has the risk of hiding an error that we are respnsible for.
Maybe it's better to do `message.downcast_ref::<Message>().unwrap()` in line `57`.

Reply:
I didn't use unwrap here because I fear that Telegram will eventually support that type and then the app would crash just because of that, but anyway in this case it would just display an empty row, so it's not great either. Maybe we should expect that a `SponsoredMessage` is only of type `MessageText` in the item_row.rs and show the 'Unsupported message' in case it's not?

Reply:
Ah ok, I think I see your point. Then, I would just leave it as it is.

Reply:
In the end I decided to both show the warning and the `MessageText` with 'Unsupported message'.